url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/sling/SlingMediator.java

```diff
@@ -0,0 +1,240 @@
+//import javax.swing.event.*;
+import java.awt.event.*;
+import java.util.Vector;
+
+import javax.swing.*;
+import javax.swing.event.*;
+
+import sjm.examples.book.imperative.*;
+import sjm.parse.Assembly;
+import sjm.parse.tokens.*;
+
+/*
+ * Copyright (c) 2000 Steven J. Metsker. All Rights Reserved.
+ */
+
+/**
+ * This class supports a SlingIde object, handling the interaction of the IDE's
+ * components.
+ * 
+ * @author Steven J. Metsker
+ * 
+ * @version 1.0
+ */
+
+public class SlingMediator implements ActionListener, ChangeListener, Runnable {
+
+	protected JButton goSource;
+	protected JButton haltSource;
+	protected JButton clearSource;
+	protected JSlider s1;
+	protected JSlider s2;
+	protected JTextArea programArea;
+	protected JTextArea messageArea;
+	protected SlingPanel plotPanel;
+
+	protected Command command;
+	protected Thread computeThread;
+
+	protected String lastProgram;
+
+	protected SlingTarget target;
+
+	protected SlingParser parser = new SlingParser();
+
+	/**
+	 * This method reacts, when the user presses one of the IDE's buttons.
+	 * 
+	 * @param ActionEvent
+	 *            the event
+	 */
+	public void actionPerformed(ActionEvent e) {
+		Object object = e.getSource();
+		if (object.equals(goSource)) {
+			go();
+		}
+		if (object.equals(clearSource)) {
+			messageArea.setText("");
+			lastProgram = null;
+		}
+		if (object == haltSource) {
+			if (computeThread != null) {
+				computeThread.stop();
+			}
+			computeThread = null;
+			setComputing(false);
+		}
+	}
+
+	/*
+	 * Throws a runtime exception if the input stream cannot be completely
+	 * recognized, and if the next element is a reserved word.
+	 */
+	protected void checkReserved(Assembly in, Assembly out) {
+		Assembly which = out == null ? in : out;
+		if (which.hasNext()) {
+			String s = which.peek().toString();
+			if (parser.wors.getReservedWords().contains(s)) {
+				throw new RuntimeException("> " + s + " is a reserved word");
+			}
+		}
+	}
+
+	/*
+	 * Throws a runtime exception if the input stream cannot be completely
+	 * recognized.
+	 */
+	protected void checkResult(String program, Assembly in, Assembly out) {
+
+		if (out == null || out.hasNext()) {
+			checkReserved(in, out);
+			if (out == null) {
+				throw new RuntimeException("> Cannot parse " + program);
+			}
+			if (out.hasNext()) {
+				throw new RuntimeException("> Input appears complete after : \n> " + out.consumed(" "));
+			}
+		}
+	}
+
+	/*
+	 * This method returns a composite command that it composes by popping
+	 * individual commands from an assembly.
+	 */
+	protected CommandSequence command(Assembly out) {
+		Vector statements = new Vector();
+		while (!out.getStack().isEmpty()) {
+			statements.addElement(out.pop());
+		}
+		CommandSequence cs = new CommandSequence();
+		int n = statements.size();
+		for (int i = n - 1; i >= 0; i--) {
+			Object o = statements.elementAt(i);
+			try {
+				cs.addCommand((Command) o);
+			} catch (ClassCastException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)